specification.md

```diff
@@ -2474,6 +2530,134 @@ authenticity of the downloaded file, for example can be the
 The exact signing and verification method is Agent specific. See
 [Code Signing](#code-signing) for recommendations.
 
+### Custom Messages
+
+#### Motivation
+
+The OpAMP protocol is intended to cover all aspects of remote management and configuration
+of Agents. If more requirements within this core functionality are identified, it is
+expected that the protocol will be extended to support this additional behavior in future
+releases. However, there are some use cases that require communication with Agents that
+will not by supported by the OpAMP protocol because they are not within the scope of Agent
+remote management. The CustomMessage allows custom behavior to be implemented between a
+Server and Agent without having to open another connection or define an entirely new
+protocol.
+
+While interoperability between many Agents and Agent Management Servers is a goal of
+OpAMP, additional custom functionality may only be supported by specific Agents and
+Servers. Where appropriate, implementers are encouraged to document their use of
+CustomMessage so that support can be added to additional platforms.
+
+#### CustomMessageCapabilities Message
+
+The Agent and Server should both use this message to signal that they support specific
+custom messages. It is supported on ServerToAgent and AgentToServer. When this message is
+sent, the list of supported custom message types should be updated to match this list. If
+this message is never sent, it is assumed that no custom message types are supported.
+
+```protobuf
+message CustomMessageCapabilities {
+  repeated string custom_message_types = 1; // Status: [Development]
+}
+```
+
+##### CustomMessageCapabilities.custom_message_types
+
+A list of custom message types that are supported. Each type should match the type
+specified in a supported CustomMessage.
+
+#### CustomMessage
+
+CustomMessage allows for custom messages to be sent between the Agent and Server. It is
+supported on ServerToAgent and AgentToServer to allow for custom communication in both
+directions. It requires that the Agent and Server both agree on the contents and encoding
+of the messages. If the message type is not recognized, the message can be ignored.
+
+A CustomMessage identifies the type of message and contains binary data with the contents
+of the message. The format of the data will depend on the message type and is outside of
+the scope of the OpAMP protocol.
+
+```protobuf
+message CustomMessage {
```

Comment:
If we think it would be helpful, how do we suggest that custom message authors implement message versioning? For a pause message, should the version be in the name with something like `io.opentelemetry.pause/v2`, or should versioning generally happen in the data field? Alternatively, should there be `version` field on `CustomMessage`? Ultimately anyone writing custom messages is free to use their preferred versioning strategy, but I think some guidelines could help encourage consistency and interoperability.

Reply:
Because the version may affect the decoding of the data field, any version information should be in the message type. I think adding `/version` to the name makes sense.

I'm going to change things slightly to add a `capability` field to CustomMessage and will document this in the name.

Reply:
I had something like that in the original proposal but I prefer to leave message IDs out of the custom message spec. If a capability requires additional request context in a response, it is free to provide that in the data payload of the response.

Reply:
If we had sequence numbers in both the `AgentToServer` and `ServerToAgent` messages, the OpAMP client/server libraries could automatically handle this to some degree for custom messages that are synchronously processed. Asynchronous handling of custom messages would still require it to be put into the message data.

Reply:
I think we should add it when we have a good use case that requires it and start with an implementation within the message and then see if we should factor that out to be used by other implementations.

Reply:
Agreed, I think that is separate from this change.